library_generation/model/config_change.py

```diff
@@ -0,0 +1,155 @@
+from library_generation.utils.proto_path_utils import find_versioned_proto_path
+from library_generation.utils.proto_path_utils import get_proto_paths
+
+
+class ChangeType(Enum):
+    GOOGLEAPIS_COMMIT = 1
+    REPO_LEVEL_CHANGE = 2
+    LIBRARIES_ADDITION = 3
+    # As of Mar. 2024, we decide not to produce this type of change because we
+    # still need to manually remove the libray.
+    # LIBRARIES_REMOVAL = 4
+    LIBRARY_LEVEL_CHANGE = 5
+    GAPIC_ADDITION = 6
+    # As of Mar. 2024, we decide not to produce this type of change because we
+    # still need to manually remove the libray.
+    # GAPIC_REMOVAL = 7
+
+
+class HashLibrary:
+    """
+    Data class to group a LibraryConfig object and its hash value together.
+    """
+
+    def __init__(self, hash_value: int, library: LibraryConfig):
+        self.hash_value = hash_value
+        self.library = library
+
+
+class LibraryChange:
+    def __init__(self, changed_param: str, latest_value: str, library_name: str = ""):
+        self.changed_param = changed_param
+        self.latest_value = latest_value
+        self.library_name = library_name
+
+
+class QualifiedCommit:
+    def __init__(self, commit: Commit, libraries: set[str]):
+        self.commit = commit
+        self.libraries = libraries
+
+
+class ConfigChange:
+    def __init__(
+        self,
+        change_to_libraries: dict[ChangeType, list[LibraryChange]],
+        baseline_config: GenerationConfig,
+        latest_config: GenerationConfig,
+    ):
+        self.change_to_libraries = change_to_libraries
+        self.baseline_config = baseline_config
+        self.latest_config = latest_config
+
+    def get_changed_libraries(self) -> Optional[list[str]]:
+        """
+        Returns library name of changed libraries.
+        None if there is a repository level change.
+        :return: library names of change libraries.
+        """
+        if ChangeType.REPO_LEVEL_CHANGE in self.change_to_libraries:
+            return None
+        library_names = []
+        for change_type, library_changes in self.change_to_libraries.items():
+            if change_type == ChangeType.GOOGLEAPIS_COMMIT:
+                library_names.extend(self.__get_library_name_from_qualified_commits())
+            else:
+                library_names.extend(
+                    [library_change.library_name for library_change in library_changes]
+                )
+        return library_names
+
+    def get_qualified_commits(
+        self,
+        repo_url: str = "https://github.com/googleapis/googleapis.git",
+    ) -> list[QualifiedCommit]:
+        """
+        Returns qualified commits from configuration change.
+
+        A qualified commit is a commit that changes at least one file (excluding
+        BUILD.bazel) within a versioned proto path in the given proto_paths.
+        :param repo_url: the repository contains the commit history.
+        :return: QualifiedCommit objects.
+        """
+        tmp_dir = "/tmp/repo"
```

Comment:
Is it possible to download googleapis in the `output` folder? This is just because we will convert our scripts to only use `output` and not `/tmp` in favor of running docker containers in parallel.

Reply:
I changed to `sh_util("get_output_folder")`